py-polars/polars/utils/udfs.py

```diff
@@ -0,0 +1,165 @@
+    """Rewrite older opcodes using py 3.11 'BINARY_OP' instead."""
+    if sys.version_info < (3, 11) and opname in _BINARY_OPCODES:
+        argrepr = _BINARY_OPCODES[opname]
+        opname = "BINARY_OP"
+    return opname, argrepr, argval
+
+
+def _is_inefficient(ops: list[tuple[str, str, Any]], apply_target: str) -> bool:
+    """Determine if bytecode indicates only simple binary ops and/or comparisons."""
+    simple_ops = _SIMPLE_FRAME_OPS if apply_target == "frame" else _SIMPLE_EXPR_OPS
+    return bool(ops) and all(op in simple_ops for op, _argrepr, _argval in ops)
+
+
+def _rewrite_as_polars_expr(
+    ops: list[tuple[str, str, Any]], col: str | None, apply_target: str
+) -> str | None:
+    """Take postfix opcode stack and translate to native polars expression."""
+    # const / unchanged
+    if len(ops) == 1:
+        return _expr(ops[0][1], col)
+    elif len(ops) >= 3:
+        if apply_target == "expr":
+            stack = []  # type: ignore[var-annotated]
+            for op in ops:
+                stack.append(
+                    op[1]
+                    if op[0].startswith("LOAD_")
+                    else (
+                        (stack.pop(), _op(*op))
+                        if op[0] == "UNARY_NOT"
+                        else (stack.pop(-2), _op(*op), stack.pop(-1))
+                    )
+                )
+            return _expr(stack[0], col)  # type: ignore[arg-type]
+
+    # TODO: frame apply (account for BINARY_SUBSCR)
+    # TODO: series apply (rewrite col expr as series)
+    return None
+
+
+def _simple_signature(function: Callable[[Any], Any]) -> bool:
+    try:
+        return len(signature(function).parameters) == 1
+    except ValueError:
+        return False
+
+
+def warn_on_inefficient_apply(
+    function: Callable[[Any], Any], columns: list[str], apply_target: str
+) -> None:
+    """Generate ``PolarsInefficientApplyWarning`` on poor usage of ``apply`` func."""
+    # only consider simple functions with a single col/param
+    if _simple_signature(function):
+        # fast function disassembly to get bytecode ops
+        ops = _get_bytecode_ops(function)
+
+        # if ops indicate a trivial function that should be native, warn about it
+        if _is_inefficient(ops, apply_target):
+            col = (columns and columns[0]) or None
+            suggestion = _rewrite_as_polars_expr(ops, col, apply_target)
+            if suggestion:
+                warnings.warn(
+                    "Expr.apply is significantly slower than the native expressions API.\n"
+                    "Only use if you absolutely CANNOT implement your logic otherwise.\n"
+                    "In this case, you can replace your function with an expression:\n"
+                    f'\033[31m-  pl.col("{columns[0]}").apply({function!r})\033[0m\n'
+                    f"\033[32m+  {suggestion}\033[0m\n",
+                    PolarsInefficientApplyWarning,
+                    stacklevel=find_stacklevel(),
```

Comment:
If you don't mind, can we think a bit on colorized output? There are instances where no color is desired (e.g. in a computing instance where stdout is not tty) or a specific color is not matching well with background (e.g. users with colored terminal background, this is especially true on vscode with custom theme).

Reply:
thanks - I took this from `black`, but they have a `--no-color` option

would a `POLARS_NO_COLORS` environment variable be acceptable here? I expect that to the majority of users, the diff-style colours would be pretty helpful

Reply:
Hmm, we should probably only colourise when we are sure we're in a suitable env for it - which we should be able to detect, rather than leave it up to the user to set a magic flag? (That they don't know exists) :)

Reply:
thanks - what should we check about `TERM`? (not something I'm familiar with)

Reply:
I've added an `in_terminal_that_supports_colour` utility function for us; turns out that this is a bit of a dark art, but I think I've covered most of the major cases...🤔